core/src/test/java/io/questdb/cutlass/pgwire/PGSecurityTest.java

```diff
@@ -0,0 +1,216 @@
+            compiler.compile("create table src (ts TIMESTAMP, name string) timestamp(ts) PARTITION BY day", sqlExecutionContext);
+            compiler.compile("snapshot prepare", sqlExecutionContext);
+            try {
+                assertQueryDisallowed("snapshot complete");
+            } finally {
+                compiler.compile("snapshot complete", sqlExecutionContext);
+            }
+        });
+    }
+
+    @Test
+    public void testDisallowSnapshotPrepare() throws Exception {
+        assertMemoryLeak(() -> {
+            compiler.compile("create table src (ts TIMESTAMP, name string) timestamp(ts) PARTITION BY day", sqlExecutionContext);
+            assertQueryDisallowed("snapshot prepare");
+        });
+    }
+
+    @Test
+    public void testDisallowsBackupDatabase() throws Exception {
+        assertMemoryLeak(() -> {
+            compiler.compile("create table src (ts TIMESTAMP, name string) timestamp(ts) PARTITION BY day", sqlExecutionContext);
+            compiler.compile("insert into src values (now(), 'foo')", sqlExecutionContext);
+            assertQueryDisallowed("backup database");
+        });
+    }
+
+    @Test
+    public void testDisallowsBackupTable() throws Exception {
+        assertMemoryLeak(() -> {
+            compiler.compile("create table src (ts TIMESTAMP, name string) timestamp(ts) PARTITION BY day", sqlExecutionContext);
+            compiler.compile("insert into src values (now(), 'foo')", sqlExecutionContext);
+            assertQueryDisallowed("backup table src");
+        });
+    }
+
+    @Test
+    @Ignore("This is failing, but repair is nop so that's ok")
+    public void testDisallowsRepairTable() throws Exception {
+        assertMemoryLeak(() -> {
+            compiler.compile("create table src (ts TIMESTAMP, name string) timestamp(ts) PARTITION BY day", sqlExecutionContext);
+            compiler.compile("insert into src values (now(), 'foo')", sqlExecutionContext);
+            assertQueryDisallowed("repair table src");
+        });
+    }
+
+    @Test
+    public void testAllowsSelect() throws Exception {
+        assertMemoryLeak(() -> {
+            compiler.compile("create table src (ts TIMESTAMP)", sqlExecutionContext);
+            assertQueryDoesNotThrowException("select * from src");
+        });
+    }
+
+    private void assertQueryDisallowed(String query) throws Exception {
+        try (
+                final PGWireServer ignored = createPGServer(READ_ONLY_CONF);
+                final Connection connection = getConnection(false, true);
+                final PreparedStatement statement = connection.prepareStatement(query)
+        ) {
+            statement.execute();
+            fail("Query '" + query + "' must fail in the read-only mode!");
+        } catch (PSQLException e) {
+            assertContains(e.getMessage(), "Write permission denied");
+        }
+    }
+
+
+    private void assertQueryDoesNotThrowException(String query) throws Exception {
```

Comment:
looking at this now I think this is a bit silly. it does not assert anything really. so perhaps it should be just `executeWithPostgres()` or something. and then `assertQueryDisallowed()` could use it too. wdyt?

Reply:
👍